### PR TITLE
vxlan: T6505: Support VXLAN VLAN-VNI range mapping in CLI

### PR DESCRIPTION
**Pull request by natali-rs1985**
## Change Summary

Now can add vlan-to-vni range mapping
```
set interfaces vxlan vxlan0 vlan-to-vni 31-35 vni 10031-10035
```
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T6505
## How to test

```
set interfaces bridge br0 member interface vxlan0
set interfaces vxlan vxlan0 parameters external
set interfaces vxlan vxlan0 source-address '127.0.0.1'
set interfaces vxlan vxlan0 vlan-to-vni 30 vni '10030'
set interfaces vxlan vxlan0 vlan-to-vni 35-40 vni '10035-10040'
commit

vyos@vyos# bridge --json vlan tunnelshow dev vxlan0
[{"ifname":"vxlan0","tunnels":[{"vlan":30,"tunid":10030},{"vlan":35,"vlanEnd":40,"tunid":10035,"tunidEnd":10040}]}]
```
## Smoketest result

```
vyos@vyos# python3 /usr/libexec/vyos/tests/smoke/cli/test_interfaces_vxlan.py
test_add_multiple_ip_addresses (__main__.VXLANInterfaceTest.test_add_multiple_ip_addresses) ... ok
test_add_single_ip_address (__main__.VXLANInterfaceTest.test_add_single_ip_address) ... ok
test_dhcp_client_options (__main__.VXLANInterfaceTest.test_dhcp_client_options) ... skipped 'not supported'
test_dhcp_disable_interface (__main__.VXLANInterfaceTest.test_dhcp_disable_interface) ... skipped 'not supported'
test_dhcp_vrf (__main__.VXLANInterfaceTest.test_dhcp_vrf) ... skipped 'not supported'
test_dhcpv6_client_options (__main__.VXLANInterfaceTest.test_dhcpv6_client_options) ... skipped 'not supported'
test_dhcpv6_vrf (__main__.VXLANInterfaceTest.test_dhcpv6_vrf) ... skipped 'not supported'
test_dhcpv6pd_auto_sla_id (__main__.VXLANInterfaceTest.test_dhcpv6pd_auto_sla_id) ... skipped 'not supported'
test_dhcpv6pd_manual_sla_id (__main__.VXLANInterfaceTest.test_dhcpv6pd_manual_sla_id) ... skipped 'not supported'
test_interface_description (__main__.VXLANInterfaceTest.test_interface_description) ... ok
test_interface_disable (__main__.VXLANInterfaceTest.test_interface_disable) ... ok
test_interface_ip_options (__main__.VXLANInterfaceTest.test_interface_ip_options) ... ok
test_interface_ipv6_options (__main__.VXLANInterfaceTest.test_interface_ipv6_options) ... ok
test_interface_mtu (__main__.VXLANInterfaceTest.test_interface_mtu) ... ok
test_ipv6_link_local_address (__main__.VXLANInterfaceTest.test_ipv6_link_local_address) ... ok
test_mtu_1200_no_ipv6_interface (__main__.VXLANInterfaceTest.test_mtu_1200_no_ipv6_interface) ... ok
test_span_mirror (__main__.VXLANInterfaceTest.test_span_mirror) ... skipped 'not supported'
test_vif_8021q_interfaces (__main__.VXLANInterfaceTest.test_vif_8021q_interfaces) ... skipped 'not supported'
test_vif_8021q_lower_up_down (__main__.VXLANInterfaceTest.test_vif_8021q_lower_up_down) ... skipped 'not supported'
test_vif_8021q_mtu_limits (__main__.VXLANInterfaceTest.test_vif_8021q_mtu_limits) ... skipped 'not supported'
test_vif_8021q_qos_change (__main__.VXLANInterfaceTest.test_vif_8021q_qos_change) ... skipped 'not supported'
test_vif_s_8021ad_vlan_interfaces (__main__.VXLANInterfaceTest.test_vif_s_8021ad_vlan_interfaces) ... skipped 'not supported'
test_vif_s_protocol_change (__main__.VXLANInterfaceTest.test_vif_s_protocol_change) ... skipped 'not supported'
test_vxlan_external (__main__.VXLANInterfaceTest.test_vxlan_external) ... ok
test_vxlan_neighbor_suppress (__main__.VXLANInterfaceTest.test_vxlan_neighbor_suppress) ... ok
test_vxlan_parameters (__main__.VXLANInterfaceTest.test_vxlan_parameters) ... ok
test_vxlan_vlan_vni_mapping (__main__.VXLANInterfaceTest.test_vxlan_vlan_vni_mapping) ... ok
test_vxlan_vni_filter (__main__.VXLANInterfaceTest.test_vxlan_vni_filter) ... ok
test_vxlan_vni_filter_add_remove (__main__.VXLANInterfaceTest.test_vxlan_vni_filter_add_remove) ... ok

----------------------------------------------------------------------
Ran 29 tests in 141.478s

OK (skipped=14)
```
## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Review by c-po — CHANGES_REQUESTED** [review]
Python validators are slow when beeing executed. The `numeric` validator was extended to also support ranges.

Example:
`<validator name="numeric" argument="--allow-range --range 0-4095"/>`

File: src/conf_mode/interfaces_vxlan.py [hunk 1]
```diff
@@ -178,13 +178,36 @@ def verify(vxlan):
...
+
+            err_msg = f'VLAN range "{vif}" does not match VNI range "{vni}"!'
+            vif_range, vni_range = list(map(int, vif.split('-'))), list(map(int, vni.split('-')))
+
+            if len(vif_range) != len(vni_range):
+                raise ConfigError(err_msg)
+
+            if len(vif_range) > 1:
+                if vni_range[0] > vni_range[-1] or vif_range[0] > vif_range[-1]:
+                    raise ConfigError('Range for VNI and VLAN ids should be increasing and not decreasing!')
```
Comment (dmbaturin) [inline src/conf_mode/interfaces_vxlan.py]: I'd find this error message confusing. I would say "the upper bound of the range must be greater than the lower bound".

**Review by dmbaturin — CHANGES_REQUESTED** [review]

File: src/conf_mode/interfaces_vxlan.py (hunk 1 above)
Comment (natali-rs1985) [reply src/conf_mode/interfaces_vxlan.py]: Done

**Comment by sever-sever** [comment]
Can you add a raiseConfigerror check for the smoketest
For example this case
```
vlans 100-105, vni 100-102
```

**Comment by natali-rs1985** [comment]
> Can you add a raiseConfigerror check for the smoketest For example this case
> 
> ```
> vlans 100-105, vni 100-102
> ```

Added raiseConfigerror checks.

File: interface-definitions/interfaces_vxlan.xml.in [hunk 2]
```diff
@@ -117,15 +117,35 @@
...
+              <valueHelp>
+                <format>&lt;start-end&gt;</format>
+                <description>VLAN ids range (use '-' as delimiter)</description>
+              </valueHelp>
               <constraint>
-                <validator name="numeric" argument="--range 0-4094"/>
+                <validator name="numeric" argument="--allow-range --range 0-4094 "/>
               </constraint>
-              <constraintErrorMessage>VLAN ID must be between 0 and 4094</constraintErrorMessage>
+              <constraintErrorMessage>Not a valid VLAN ID or range</constraintErrorMessage>
```
Comment (c-po) [inline interface-definitions/interfaces_vxlan.xml.in]: Please change to: `Not a valid VLAN ID or range, VLAN ID must be between 0 and 4094` to give a proper hint to the user.

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/interfaces_vxlan.xml.in (hunk 2 above)
Comment (c-po) [reply interface-definitions/interfaces_vxlan.xml.in]: Other then that it looks good to me.

**Review by c-po — APPROVED** [review]

File: interface-definitions/interfaces_vxlan.xml.in [hunk 3]
```diff
@@ -117,15 +117,35 @@
                 <format>u32:0-4094</format>
                 <description>Virtual Local Area Network (VLAN) ID</description>
               </valueHelp>
+              <valueHelp>
+                <format>&lt;start-end&gt;</format>
+                <description>VLAN ids range (use '-' as delimiter)</description>
```
Comment (dmbaturin) [inline interface-definitions/interfaces_vxlan.xml.in]: Let's keep the "ID" spelling consistently capitalized.

**Review by dmbaturin — APPROVED** [review]
I see no issues with the implementation, just would prefer a minor spelling change.

File: interface-definitions/interfaces_vxlan.xml.in (hunk 3 above)
Comment (natali-rs1985) [reply interface-definitions/interfaces_vxlan.xml.in]: Done

**Review by sever-sever — APPROVED** [review]

**Comment by sever-sever** [comment]
@MergifyIo backport circinus sagitta

**Pull request merged**